miner/src/pool/listener.rs

```diff
@@ -116,6 +117,69 @@ impl txpool::Listener<Transaction> for Logger {
 	}
 }
 
+/// Transactions pool notifier
+#[derive(Default)]
+pub struct TransactionsPoolNotifier {
+	listeners: Vec<Mutex<mpsc::Sender<(H256, TxStatus)>>>,
+}
+
+impl TransactionsPoolNotifier {
+	/// Add new listener to receive notifications.
+	pub fn add(&mut self, f: mpsc::Sender<(H256, TxStatus)>) {
+		self.listeners.push(Mutex::new(f));
+	}
+
+	/// Notify listeners about all currently transactions.
+	fn notify(& mut self, hash: H256, status: TxStatus) {
+		for l in &self.listeners {
+			let l = l.lock();
+			if let Ok(l) = l {
+				let _ = l.send((hash, status.clone()));
+			}
+		}
+	}
+}
+
+impl fmt::Debug for TransactionsPoolNotifier {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		fmt.debug_struct("TransactionsPoolNotifier")
+			.field("listeners", &self.listeners.len())
+			.finish()
+	}
+}
+
+impl txpool::Listener<Transaction> for TransactionsPoolNotifier {
+	fn added(&mut self, tx: &Arc<Transaction>, _old: Option<&Arc<Transaction>>) {
+		let hash = tx.hash.clone();
+		self.notify(hash, TxStatus::Added);
+	}
+
+	fn rejected(&mut self, tx: &Arc<Transaction>, _reason: &txpool::ErrorKind) {
```

Comment:
I just realised that this might be called really frequently, and currently we need to iterate over all senders and have a unique lock for this, so this will have drastic performance consequences.

Instead I think we should do something very similar that we do with `Notifer` in the same file, i.e:
1. We first aggregate all the notifications that should be sent
2. Then we send them all at once after batch import to the queue is finished.

It might be even worth to remove `Notifier` and convert this other notification to use this (channel-based) notifier instead.

```rust
struct Notifier {
  pending: HashMap<H256, TxStatus>,
  listeners: Vec<futures::sync::mpsc::UnboundedSender<Arc<HashMap<H256, TxStatus>>>>,
}

impl Notifier {
 fn notify(&mut self) {
   let to_send = Arc::new(std::mem::replace(&mut self.pending, HashMap::new));
   self.listeners.retain(|listener| listener.unbounded_send(to_send.clone()).is_ok());
 }
}

impl txpool::Listener<Transaction> for TransactionPoolNotifier {
  fn added(&mut self, ...) {
    self.pending.insert(tx.hash, TxStatus::Added);
  }
  ...
}
```

Reply:
@IntegralTeam Please let me know if you need any more explanations or help with this.

Reply:
@tomusdrw Thank you, I've got what I need to do.

But, I think, it's not quite right to use `HashMap` here, because transaction firstly could be added with according `Added` status, and then before sending notification, same transaction could get status `Culled` and it will lead that `Added` status will be erased from `HashMap`

Maybe it will be better to use tuple vector `Vec<(H256, TxStatus)>`?

Reply:
@IntegralTeam good point, yeah if we never want to miss any event and have them in order, then `Vec` is indeed a good choice here.

Reply:
Ok, let's do it this way.